### PR TITLE
KAFKA-10287: Skip unknown offsets when computing sum of changelog offsets

### PR DESCRIPTION
**Pull request by cadonna**
In PR #8962 we introduced a sentinel UNKNOWN_OFFSET to mark unknown offsets in checkpoint files. The sentinel was set to -2 which is the same value used for the sentinel LATEST_OFFSET that is used in subscriptions to signal that state stores have been used by an active task. Unfortunately, we missed to skip UNKNOWN_OFFSET when we compute the sum of the changelog offsets. 

If a task had only one state store and it did not restore anything before the next rebalance, the stream thread wrote -2 (i.e., UNKNOWN_OFFSET) into the subscription as sum of the changelog offsets. During assignment, the leader interpreted the -2 as if the stream run the task as active although it might have run it as standby. This misinterpretation of the sentinel value resulted in unexpected task assigments.  

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by cadonna** [comment]
Started 10x runs of the failing system test here: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4074/

**Comment by cadonna** [comment]
Call for review: @ableegoldman @mjsax @vvcephei @guozhangwang

**Comment by mjsax** [comment]
Retest this please.

**Review by mjsax — APPROVED** [review]

**Comment by mjsax** [comment]
@cadonna Should the PR be against `trunk` ?

**Review by ableegoldman — COMMENTED** [review]
Aw man, can't believe I missed that...I suppose this would mean KIP-441 is slightly broken for any subtopology with in-memory stores? I notice in `TaskManager#sumOfChangelogOffsets` we return early if we find any offset that's equal to `LATEST_OFFSET`. 

Maybe as extra protection we should not do that, and make sure all offsets are `LATEST_OFFSET` before assuming the task is actually caught up? But I'm fine doing that in a follow PR so we can get this merged ASAP -- seems worth asking Randall if we can slip this fix into 2.6, especially if a new RC has to be cut regardless (I think there was one outstanding issue last I saw).

If this is causing system tests to fail then it should be an automatic blocker anyway?

**Comment by cadonna** [comment]
I started a run of Streams system tests:
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4077/

**Comment by cadonna** [comment]
@mjsax I thought this fix is most urgently needed on 2.6 to unblock the release. I will open a new PR for trunk.

**Comment by cadonna** [comment]
@ableegoldman I agree it should be a blocker.

**Comment by cadonna** [comment]
@ableegoldman I agree that having the extra protection you propose makes sense. I will open a follow-up PR on trunk.

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
Ok to test

**Comment by vvcephei** [comment]
Test this please

**Review by chia7712 — APPROVED** [review]

**Review by ableegoldman — APPROVED** [review]
Thanks Bruno. I do agree with Matthias that we should target this against trunk and then cherrypick back, but it's not worth changing now if it will delay the release more

**Review by vvcephei — COMMENTED** [review]
Hey @cadonna , thanks for the fix and tests!

I took a look around the codebase, and found that we already have:
`public static final long UNKNOWN_OFFSET_SUM = -3L` in `SubscriptionInfo`. Is there any possible conflict there?

**Comment by ableegoldman** [comment]
@vvcephei I don't think there should be any issue there, since the checkpointed  `OFFSET_UNKNOWN` is just ignored on read (or should be) and will never make it into the subscription info.

That said, it does make me a bit nervous and I wouldn't count on this always being true, so I'd rather choose a different sentinel to be safe. 

Maybe we can put together a brief list of all sentinel values used in Streams (and the clients) to keep track and avoid issues like this going forward. I had to spend a long time digging around the client code to figure out whether an offset of -1 was ever possible in non-error cases (the answer seems to be "yes"). We'd have to actually keep it up to date of course, which might be...challenging

**Comment by cadonna** [comment]
@vvcephei Thanks for checking! Given your finding, I will change `UNKNOWN_OFFSET` to `-4L`. Actually, we do not to change `UNKNOWN_OFFSET` at all, because we now never write it into the offset sums. I just thought it would be better to change it to not end up in a similar situation after a future refactoring. I will follow the same reasoning in this case.

**Review by vvcephei — APPROVED** [review]
Thanks, @cadonna !

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
Test this please

**Comment by chia7712** [comment]
Test this please

**Comment by cadonna** [comment]
Java 11 failed with

```
08:16:38 org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest > shouldUpgradeFromEosAlphaToEosBeta[true] FAILED
08:16:38     java.lang.AssertionError: Did not receive all 20 records from topic multiPartitionOutputTopic within 60000 ms
08:16:38     Expected: is a value equal to or greater than <20>
08:16:38          but: <0> was less than <20>
```

Java 14 failed with 

```

06:37:03 kafka.network.SocketServerTest > testIdleConnection FAILED
06:37:03     org.scalatest.exceptions.TestFailedException: Failed to close idle channel

```

Java 8 failed with

```
org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[true]

java.lang.AssertionError: Did not receive all 168 records from topic multiPartitionOutputTopic within 60000 ms
Expected: is a value equal to or greater than <168>
     but: <167> was less than <168>
```

**Comment by chia7712** [comment]
```kafka.network.SocketServerTest``` and ```org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest``` works on my local.

**Comment by chia7712** [comment]
Test this please

**Comment by mjsax** [comment]
@chia7712 Only committers can trigger Jenkins builds.

**Comment by mjsax** [comment]
Java 8 passed the tests in question are know to be flaky. Might be good enough? Leave it to @vvcephei to make a call and merge.

**Comment by chia7712** [comment]
> Only committers can trigger Jenkins builds.

It seems I can trigger the jenkins by comment “ Test this please”. Let me test it again.

**Comment by chia7712** [comment]
Jenkins is running now :(

Maybe it does not exclude the committers from other Apache projects. At any rate, I will not trigger Jenkins in the future.

**Comment by mjsax** [comment]
Ah yes. I work for all committers -- it's not project specific. Missed that you are HBase Committer/PMC :)

**Comment by cadonna** [comment]
Java 8 was aborted due to a timeout.

```
04:56:02 Build timed out (after 270 minutes). Marking the build as aborted.
```

**Comment by cadonna** [comment]
@chia7712 Please do trigger Jenkins in the future. The more we have that have permission to trigger Jenkins and do it, the better it is. Especially for people like me that do not have permissions to trigger it.

**Comment by chia7712** [comment]
> Please do trigger Jenkins in the future.

Copy that. Let me do it at once. Test this please

**Comment by cadonna** [comment]
Java 11 failed with

```
org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest > shouldUpgradeFromEosAlphaToEosBeta[false]
13:25:38     Expected: <[KeyValue(0, 0), KeyValue(0, 1), KeyValue(0, 3), KeyValue(0, 6), KeyValue(0, 10), KeyValue(0, 15), KeyValue(0, 21), KeyValue(0, 28), KeyValue(0, 36), KeyValue(0, 45), KeyValue(0, 55), KeyValue(0, 66), KeyValue(0, 78), KeyValue(0, 91), KeyValue(0, 105)]>
13:25:38          but: was <[KeyValue(0, 0), KeyValue(0, 0), KeyValue(0, 1), KeyValue(0, 3), KeyValue(0, 6), KeyValue(0, 10), KeyValue(0, 15), KeyValue(0, 21), KeyValue(0, 28), KeyValue(0, 36), KeyValue(0, 45), KeyValue(0, 55)]>

13:29:44 org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest > shouldUpgradeFromEosAlphaToEosBeta[true] FAILED
13:29:44     java.lang.AssertionError: Did not receive all 168 records from topic multiPartitionOutputTopic within 60000 ms
13:29:44     Expected: is a value equal to or greater than <168>
13:29:44          but: <167> was less than <168>

13:47:10 org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest > testReplication FAILED
13:47:10     java.lang.AssertionError: Connector MirrorCheckpointConnector tasks did not start in time on cluster: org.apache.kafka.connect.util.clusters.EmbeddedConnectCluster@7e44a25d
13:47:10         at org.apache.kafka.connect.util.clusters.EmbeddedConnectClusterAssertions.assertConnectorAndAtLeastNumTasksAreRunning(EmbeddedConnectClusterAssertions.java:296)
13:47:10         at org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.waitUntilMirrorMakerIsRunning(MirrorConnectorsIntegrationTest.java:191)
13:47:10         at org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.setup(MirrorConnectorsIntegrationTest.java:184)
13:47:10 
13:47:10         Caused by:
13:47:10         java.lang.AssertionError: Condition not met within timeout 30000. The connector or at least 1 of tasks are not running.
```

Java 14 failed with

```
13:22:42 > Task :connect:mirror:integrationTest
13:22:42 
13:22:42 org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest > testReplication STARTED
13:24:09 
13:24:09 org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest > testReplication SKIPPED
13:24:09 
13:24:09 > Task :connect:mirror:integrationTest FAILED
```

**Comment by chia7712** [comment]
Test this please

**Comment by rhauch** [comment]
@chia7712, looks like all tests passed.

**Comment by chia7712** [comment]
> looks like all tests passed.

yep. we need someone to merge it. @vvcephei @mjsax :)

**Comment by rhauch** [comment]
@vvcephei, @mjsax: this is targeted to the `2.6` branch. Does this need to be forward merged to `trunk`, or is that the purpose of #9074?

**Comment by vvcephei** [comment]
Thanks all. Merged to 2.6.

@cadonna will follow up with a fresh fix for trunk.

**Comment by vvcephei** [comment]
Note, I've tracked the need for a trunk port with https://issues.apache.org/jira/browse/KAFKA-10319

**Pull request merged**
